datafusion/functions-array/src/range.rs

```diff
@@ -0,0 +1,332 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! [`ScalarUDFImpl`] definitions for range and gen_series functions.
+
+use arrow::array::{Array, ArrayRef, Int64Array, ListArray};
+use arrow::datatypes::{DataType, Field};
+use arrow_buffer::{BooleanBufferBuilder, NullBuffer, OffsetBuffer};
+use std::any::Any;
+
+use arrow_array::types::{Date32Type, IntervalMonthDayNanoType};
+use arrow_array::Date32Array;
+use arrow_schema::IntervalUnit::MonthDayNano;
+use datafusion_common::cast::{as_date32_array, as_int64_array, as_interval_mdn_array};
+use datafusion_common::{exec_err, not_impl_datafusion_err, Result};
+use datafusion_expr::expr::ScalarFunction;
+use datafusion_expr::Expr;
+use datafusion_expr::{
+    ColumnarValue, ScalarUDFImpl, Signature, TypeSignature, Volatility,
+};
+use std::sync::Arc;
+
+make_udf_function!(
+    Range,
+    range,
+    start stop step,
+    "create a list of values in the range between start and stop",
+    range_udf
+);
+#[derive(Debug)]
+pub(super) struct Range {
+    signature: Signature,
+    aliases: Vec<String>,
+}
+impl Range {
+    pub fn new() -> Self {
+        use DataType::*;
+        Self {
+            signature: Signature::one_of(
+                vec![
+                    TypeSignature::Exact(vec![Int64]),
+                    TypeSignature::Exact(vec![Int64, Int64]),
+                    TypeSignature::Exact(vec![Int64, Int64, Int64]),
+                    TypeSignature::Exact(vec![Date32, Date32, Interval(MonthDayNano)]),
+                ],
+                Volatility::Immutable,
+            ),
+            aliases: vec![String::from("range")],
+        }
+    }
+}
+impl ScalarUDFImpl for Range {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn name(&self) -> &str {
+        "range"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        use DataType::*;
+        Ok(List(Arc::new(Field::new(
+            "item",
+            arg_types[0].clone(),
+            true,
+        ))))
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        let args = ColumnarValue::values_to_arrays(args)?;
```

Comment:
should we change with make_scalar_function wrapper

Reply:
`gen_range_inner` and `gen_range_date` functions expect 2 parameters where `make_scalar_function` expects 1 parameter when calling these functions so this causes compile-time issue.

Reply:
@erenavsarogullari 
You can try wrapping it with closure if the args are not consistent.
```rust
        match args[0].data_type() {
            DataType::Int64 => make_scalar_function(|args| gen_range_inner(args, false))(args),
            DataType::Date32 => make_scalar_function(|args| gen_range_date(&args, false))(args),
            _ => {
                exec_err!("unsupported type for range")
            }
        }
```